util/extsort/disk_sorter.go

```diff
@@ -15,37 +15,654 @@
-	tmpDir string        // directory for temporary files
-	idGen  *atomic.Int64 // id generator for sst files
+func (si *sstIter) Close() error {
+	err := si.iter.Close()
+	if si.onClose != nil {
+		err = goerrors.Join(err, si.onClose())
+	}
+	return errors.Trace(err)
+}
+
+type mergingIter struct {
+	heap mergingIterHeap
+	// orderedFiles is the files to be merged. It is ordered by start key.
+	orderedFiles []*fileMetadata
+	// nextFileIndex is the index of the next file to be read.
+	// mergingIter only maintains minimum set of opened files.
+	// If current key is less than the start key of next file,
+	// the next file does not need to be opened.
+	nextFileIndex int
+	openIter      openIterFunc
+	curKey        []byte // only used in Next() to avoid alloc
+	err           error
+}
+
+type mergingIterItem struct {
+	Iterator
+	index int // index in orderedFiles
+}
+
+type mergingIterHeap []mergingIterItem
+
+func (h mergingIterHeap) Len() int {
+	return len(h)
+}
+
+func (h mergingIterHeap) Swap(i, j int) {
+	h[i], h[j] = h[j], h[i]
+}
+
+func (h mergingIterHeap) Less(i, j int) bool {
+	return bytes.Compare(h[i].UnsafeKey(), h[j].UnsafeKey()) < 0
+}
+
+func (h *mergingIterHeap) Push(x interface{}) {
+	*h = append(*h, x.(mergingIterItem))
+}
+
+func (h *mergingIterHeap) Pop() interface{} {
+	old := *h
+	n := len(old)
+	x := old[n-1]
+	*h = old[:n-1]
+	return x
+}
+
+type openIterFunc func(file *fileMetadata) (Iterator, error)
+
+// newMergingIter returns an iterator that merges the given files.
+// orderedFiles must be ordered by start key, otherwise it panics.
+func newMergingIter(orderedFiles []*fileMetadata, openIter openIterFunc) *mergingIter {
+	if !slices.IsSortedFunc(orderedFiles, func(a, b *fileMetadata) bool {
+		return bytes.Compare(a.startKey, b.startKey) < 0
+	}) {
+		panic("newMergingIter: orderedFiles are not ordered by start key")
+	}
+	return &mergingIter{
+		orderedFiles: orderedFiles,
+		openIter:     openIter,
+	}
+}
+
+func (m *mergingIter) Seek(key []byte) bool {
+	m.err = nil
+	oldHeap := m.heap
+	m.heap = nil
+
+	var toClose []mergingIterItem
+	openedIters := make(map[int]struct{})
+	for i, item := range oldHeap {
+		file := m.orderedFiles[item.index]
+		if bytes.Compare(key, file.startKey) >= 0 &&
+			bytes.Compare(key, file.endKey) < 0 &&
+			item.Seek(key) {
+			openedIters[item.index] = struct{}{}
+			heap.Push(&m.heap, item)
+		} else {
+			m.err = item.Error()
+			if m.err != nil {
+				// Iterators in m.heap should be closed by mergingIter.Close().
+				toClose = append(toClose, oldHeap[i:]...)
+				break
+			}
+			// The seek key is not in the range of [startKey, endKey), close it.
+			toClose = append(toClose, item)
+		}
+	}
+	for _, item := range toClose {
+		if err := item.Close(); err != nil && m.err == nil {
+			m.err = err
+		}
+	}
+	if m.err != nil {
+		return false
+	}
+
+	m.nextFileIndex = len(m.orderedFiles)
+	for i, file := range m.orderedFiles {
+		if bytes.Compare(file.startKey, key) > 0 {
```

Comment:
`>=` ?

Reply:
no

Reply:
suppose we have 2 files, and we want to seek to `a`, suppose it's the first time we seek, that means heap is empty, so the first file is skpped
```
a...f
 b.....g
```

Reply:
Why? For the first file, `bytes.Compare(file.startKey, key)  > 0` should return false.

Reply:
you're right, i misunderstand